File: app/src/main/java/kr/gyeongmunbaek/imagesequenceanimaion/view/ImageSequenceAnimationView.java
package kr.gyeongmunbaek.imagesequenceanimaion.view;

import android.content.Context;
import android.content.res.Resources;
import android.graphics.Bitmap;
import android.graphics.Canvas;
import android.graphics.RectF;
import android.os.AsyncTask;
import android.os.Handler;
import android.os.Message;
import android.util.AttributeSet;
import android.view.View;

import kr.gyeongmunbaek.imagesequenceanimaion.view.image.BitmapCreatorParent;
import kr.gyeongmunbaek.imagesequenceanimaion.view.image.cache.CustomCache;

/**
 * Created by gyeongmunbaek on 2017-10-26.
 */

public class ImageSequenceAnimationView extends View {
    private static final int CACHE_COUNT = 5;
    private static final int MESSAGE_INVALIDATE = 1;
    private static final int MESSAGE_START_ANIMATION = 2;

    private static final int MESSAGE_DELAY = 10;

    private final Context mContext;

    private CustomCache mCustomCache = null;

    private int mCacheCount = CACHE_COUNT;
    private int mDelay = MESSAGE_DELAY;

    private BitmapCreatorParent mBitmapClass = null;

    private int mCurrentIndex = 0;
    private Bitmap mPreBitmap = null;
    private boolean mIsPlaying = false;

    private Handler mHandler = new Handler(new Handler.Callback() {

        @Override
        public boolean handleMessage(Message msg) {
            if (msg.what == MESSAGE_INVALIDATE) {
                new Runnable() {
                    @Override
                    public void run() {
                        if (mIsPlaying) {
                            invalidate();
                            update(mCurrentIndex);
                        }
                    }
                }.run();
            } else if (msg.what == MESSAGE_START_ANIMATION) {
                mHandler.removeMessages(MESSAGE_INVALIDATE);
                new Runnable() {
                    @Override
                    public void run() {
                        setBitmapClass(mBitmapClass);
                    }
                }.run();
            }
            return false;
        }
    });

    public ImageSequenceAnimationView(Context context) {
        super(context);
        mContext = context;
    }

    public ImageSequenceAnimationView(Context context, AttributeSet attrs) {
        super(context, attrs);
        mContext = context;
    }

    public ImageSequenceAnimationView(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        mContext = context;
    }

    public ImageSequenceAnimationView(Context context, AttributeSet attrs, int defStyleAttr, int defStyleRes) {
        super(context, attrs, defStyleAttr, defStyleRes);
        mContext = context;
    }

    @Override
    protected void onDraw(Canvas canvas) {
        if (canvas.getWidth() > 0 && canvas.getHeight() > 0) {
            if (mCustomCache != null && mCustomCache.get(mCurrentIndex) != null) {
                Bitmap lBitmap = mCustomCache.get(mCurrentIndex).getBitmap();

                if (lBitmap != null) {
                    mPreBitmap = lBitmap;
                    canvas.drawBitmap(
                            lBitmap,
                            null,
                            new RectF(0, 0, canvas.getWidth(), canvas.getHeight()),
                            null);

                    mCurrentIndex++;
                } else if (mPreBitmap != null && mPreBitmap.isRecycled() == false) {
                    canvas.drawBitmap(
                            mPreBitmap,
                            null,
                            new RectF(0, 0, canvas.getWidth(), canvas.getHeight()),
                            null);
                }
            }
        }
    }

    public ImageSequenceAnimationView setCacheCount(int pCount) {
        mCacheCount = pCount;
        return this;
    }

    public ImageSequenceAnimationView setDelay(int pDelay) {
        mDelay = pDelay;
        return this;
    }

    public ImageSequenceAnimationView setAnimationClazz(final BitmapCreatorParent pClass) throws NullPointerException {
        if (pClass == null) {
            throw new NullPointerException();
        }
        mBitmapClass = pClass;
        return this;
    }

    public void startAnimation() {
        if (mHandler != null) {
            mHandler.removeMessages(MESSAGE_START_ANIMATION);
            Message msg = mHandler.obtainMessage(MESSAGE_START_ANIMATION);
            mHandler.sendMessageDelayed(msg, mDelay);
        }
    }

    private void setBitmapClass(final BitmapCreatorParent pClass) {
        new AsyncTask<Void, Void, Void>() {
            @Override
            protected Void doInBackground(Void... params) {
                mIsPlaying = false;

                if (mHandler != null) {
                    mHandler.removeMessages(MESSAGE_INVALIDATE);
                    mHandler.removeMessages(MESSAGE_START_ANIMATION);
                }

                if (mCustomCache != null) {
                    mCustomCache.evictAll();
                    mCustomCache = null;
                }

                if (mPreBitmap != null) {
                    if (mPreBitmap.isRecycled() == false) {
                        mPreBitmap.recycle();
                    }
                    mPreBitmap = null;
                }

                mCustomCache = new CustomCache(mContext, mCacheCount, pClass);
                mCurrentIndex = 0;

                for (int index = 0; index < mCacheCount; index++) {
                    try {
                        mCustomCache.get(index).createBitmap();
                    } catch (OutOfMemoryError e) {

                    } catch (Resources.NotFoundException e1) {

                    }
                }

                mIsPlaying = true;
                if (mHandler != null) {
                    Message msg = mHandler.obtainMessage(MESSAGE_INVALIDATE);
                    mHandler.sendMessageDelayed(msg, mDelay);
                }
                return null;
            }
        }.execute();
    }

    public void destroy() {
        mIsPlaying = false;

        if (mHandler != null) {
            mHandler.removeMessages(MESSAGE_INVALIDATE);
            mHandler.removeMessages(MESSAGE_START_ANIMATION);
        }

        if (mCustomCache != null) {
            mCustomCache.evictAll();
            mCustomCache = null;
        }

        if (mPreBitmap != null) {
            if (mPreBitmap.isRecycled() == false) {
                mPreBitmap.recycle();
            }
            mPreBitmap = null;
        }
    }

    @Override
    protected void onDetachedFromWindow() {
        super.onDetachedFromWindow();

        destroy();
    }

    private void update(final int pCurrentIndex) {
        new AsyncTask<Void, Void, Void>() {
            @Override
            protected Void doInBackground(Void... params) {
                if (mIsPlaying) {
                    if (mCustomCache != null) {
                        for (int index = pCurrentIndex; index < pCurrentIndex + CACHE_COUNT; index++) {
                            if (mCustomCache != null && mCustomCache.get(index) != null) {
                                try {
                                    mCustomCache.get(index).createBitmap();
                                } catch (OutOfMemoryError e) {

                                } catch (Resources.NotFoundException e1) {

                                }
                            }
                        }
                    }
                    if (mHandler != null) {
                        mHandler.removeMessages(MESSAGE_INVALIDATE);
                        Message msg = mHandler.obtainMessage(MESSAGE_INVALIDATE);
                        mHandler.sendMessageDelayed(msg, mDelay);
                    }
                }
                return null;
            }
        }.execute();
    }
}

File: app/src/main/java/kr/gyeongmunbaek/imagesequenceanimaion/view/ImageSequenceAnimationView.kt
package kr.gyeongmunbaek.imagesequenceanimaion.view

import android.content.Context
import android.content.res.Resources
import android.graphics.Bitmap
import android.graphics.Canvas
import android.graphics.RectF
import android.os.AsyncTask
import android.os.Handler
import android.os.Message
import android.util.AttributeSet
import android.view.View

import kr.gyeongmunbaek.imagesequenceanimaion.view.image.BitmapCreatorParent
import kr.gyeongmunbaek.imagesequenceanimaion.view.image.cache.CustomCache

/**
 * Created by gyeongmunbaek on 2017-10-26.
 */

class ImageSequenceAnimationView : View {
    private var customCache: CustomCache<BitmapCreatorParent>? = null
    private var cacheCount = CACHE_COUNT
    private var delay = MESSAGE_DELAY

    private var bitmapClass: BitmapCreatorParent? = null

    private var currentIndex = 0
    private var preBitmap: Bitmap? = null
    private var isPlaying = false

    private var area: RectF = RectF(0.0f, 0.0f, width.toFloat(), height.toFloat())
    private val handler = CustomHandler(this)

    constructor(context: Context) : super(context)

    constructor(context: Context, attrs: AttributeSet) : super(context, attrs)

    constructor(context: Context, attrs: AttributeSet, defStyleAttr: Int) : super(context, attrs, defStyleAttr)

    constructor(context: Context, attrs: AttributeSet, defStyleAttr: Int, defStyleRes: Int) : super(context, attrs, defStyleAttr, defStyleRes)

    override fun onDraw(canvas: Canvas) {
        if (width > 0 && height > 0) {
            if (customCache != null && customCache!![currentIndex] != null) {
                customCache!![currentIndex]?.bitmap?.let {
                    preBitmap = it
                    canvas.drawBitmap(
                            it, null,
                            area, null)

                    currentIndex++
                } ?: run {
                    preBitmap?.let { bitmap ->
                        if (!bitmap.isRecycled) {
                            canvas.drawBitmap(
                                    bitmap, null,
                                    area, null)
                        }
                    }
                }
            }
        }
    }

    fun setCacheCount(pCount: Int): ImageSequenceAnimationView {
        cacheCount = pCount
        return this
    }

    fun setDelay(delay: Long): ImageSequenceAnimationView {
        this.delay = delay
        return this
    }

    @Throws(NullPointerException::class)
    fun setAnimationClazz(pClass: BitmapCreatorParent?): ImageSequenceAnimationView {
        if (pClass == null) {
            throw NullPointerException()
        }
        bitmapClass = pClass
        return this
    }

    fun startAnimation() {
        handler.removeMessages(MESSAGE_START_ANIMATION)
        val msg = handler.obtainMessage(MESSAGE_START_ANIMATION)
        handler.sendMessageDelayed(msg, delay)
    }

    private fun setBitmapClass(bitmapCreatorParent: BitmapCreatorParent?) {
        val bitmapTask = BitmapClassAsyncTask(this, bitmapCreatorParent)
        bitmapTask.execute()
    }

    fun destroy() {
        isPlaying = false

        handler.removeMessages(MESSAGE_INVALIDATE)
        handler.removeMessages(MESSAGE_START_ANIMATION)

        if (customCache != null) {
            customCache!!.evictAll()
            customCache = null
        }

        if (preBitmap != null) {
            if (!preBitmap!!.isRecycled) {
                preBitmap!!.recycle()
            }
            preBitmap = null
        }
    }

    override fun onDetachedFromWindow() {
        super.onDetachedFromWindow()

        destroy()
    }

    private fun update(currentIndex: Int) {
        val asyncTask = UpdateAsyncTask(this, currentIndex)
        asyncTask.execute()
    }

    companion object {
        private const val CACHE_COUNT = 5
        private const val MESSAGE_INVALIDATE = 1
        private const val MESSAGE_START_ANIMATION = 2

        private const val MESSAGE_DELAY = 10L
    }

    class CustomHandler(private val view: ImageSequenceAnimationView) : Handler() {
        override fun handleMessage(msg: Message) {
            view.run {
                if (msg.what == MESSAGE_INVALIDATE) {
                    Runnable {
                        if (isPlaying) {
                            area = RectF(0.0f, 0.0f, width.toFloat(), height.toFloat())
                            invalidate()
                            update(currentIndex)
                        }
                    }.run()
                } else if (msg.what == MESSAGE_START_ANIMATION) {
                    handler.removeMessages(MESSAGE_INVALIDATE)
                    Runnable { setBitmapClass(bitmapClass) }.run()
                }
            }
        }
    }

    class BitmapClassAsyncTask(private val view: ImageSequenceAnimationView, var bitmapCreatorParent: BitmapCreatorParent?) : AsyncTask<Void, Void, Void>() {
        override fun doInBackground(vararg params: Void?): Void? {
            view.run {
                isPlaying = false

                handler.removeMessages(MESSAGE_INVALIDATE)
                handler.removeMessages(MESSAGE_START_ANIMATION)

                if (customCache != null) {
                    customCache!!.evictAll()
                    customCache = null
                }

                preBitmap?.let {
                    if (!it.isRecycled) {
                        it.recycle()
                    }
                    preBitmap = null
                }

                bitmapCreatorParent?.let {
                    customCache = CustomCache(context, cacheCount, it)
                    currentIndex = 0
                }

                for (index in 0 until cacheCount) {
                    try {
                        customCache!![index]?.createBitmap()
                    } catch (e: OutOfMemoryError) {

                    } catch (e1: Resources.NotFoundException) {

                    }
                }

                isPlaying = true
                val msg = handler.obtainMessage(MESSAGE_INVALIDATE)
                handler.sendMessageDelayed(msg, delay)
                return null
            }
        }
    }

    class UpdateAsyncTask(private val view: ImageSequenceAnimationView, private val index: Int) : AsyncTask<Void, Void, Void>() {
        override fun doInBackground(vararg params: Void): Void? {
            view.run {
                if (isPlaying) {
                    if (customCache != null) {
                        for (index in index until index + CACHE_COUNT) {
                            if (customCache != null && customCache!![index] != null) {
                                try {
                                    customCache!![index]?.createBitmap()
                                } catch (e: OutOfMemoryError) {

                                } catch (e1: Resources.NotFoundException) {

                                }
                            }
                        }
                    }
                    handler.removeMessages(MESSAGE_INVALIDATE)
                    val msg = handler.obtainMessage(MESSAGE_INVALIDATE)
                    handler.sendMessageDelayed(msg, delay)
                }
                return null
            }
        }
    }
}
